src/specular/stages/batcher_transactions.rs

```diff
@@ -0,0 +1,101 @@
+use std::sync::mpsc;
+
+use ethers::{
+    abi::parse_abi_str,
+    contract::Lazy,
+    prelude::BaseContract,
+    types::{Bytes, Selector, U256},
+};
+use eyre::Result;
+use std::collections::VecDeque;
+
+use crate::derive::stages::batcher_transactions::BatcherTransactionMessage;
+use crate::derive::PurgeableIterator;
+
+pub struct SpecularBatcherTransactions {
+    txs: VecDeque<SpecularBatcherTransaction>,
+    transaction_rx: mpsc::Receiver<BatcherTransactionMessage>,
+}
+
+impl Iterator for SpecularBatcherTransactions {
+    type Item = SpecularBatcherTransaction;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.process_incoming();
+        self.txs.pop_front()
+    }
+}
+
+impl PurgeableIterator for SpecularBatcherTransactions {
+    fn purge(&mut self) {
+        // drain the channel first
+        while self.transaction_rx.try_recv().is_ok() {}
+        self.txs.clear();
+    }
+}
+
+impl SpecularBatcherTransactions {
+    pub fn new(transaction_rx: mpsc::Receiver<BatcherTransactionMessage>) -> Self {
+        Self {
+            transaction_rx,
+            txs: VecDeque::new(),
+        }
+    }
+
+    pub fn process_incoming(&mut self) {
+        while let Ok(BatcherTransactionMessage { txs, l1_origin }) = self.transaction_rx.try_recv()
+        {
+            for data in txs {
+                let res = SpecularBatcherTransaction::new(l1_origin, &data).map(|tx| {
+                    self.txs.push_back(tx);
+                });
+
+                if res.is_err() {
+                    tracing::warn!("dropping invalid batcher transaction");
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SpecularBatcherTransaction {
+    pub l1_inclusion_block: u64,
+    pub version: u64,
+    pub tx_batch: Bytes,
+}
+
+impl SpecularBatcherTransaction {
+    pub fn new(l1_inclusion_block: u64, data: &[u8]) -> Result<Self> {
+        if data.len() < 4 {
+            eyre::bail!("invalid transaction data");
+        }
+        if data[..4] != *APPEND_TX_BATCH_SELECTOR {
+            eyre::bail!("not appendTxBatch call");
+        }
```

Comment:
i'd move these checks to `process_incoming` and just filter them out rather than raising an error